src/main/java/fi/aalto/cs/apluscourses/intellij/actions/ImportProjectSettingsAction.java

```diff
@@ -0,0 +1,133 @@
+import net.lingala.zip4j.model.FileHeader;
+import org.jetbrains.annotations.NotNull;
+
+public class ImportProjectSettingsAction extends AnAction {
+  @NotNull
+  private MainViewModelProvider mainViewModelProvider;
+
+  @NotNull ProjectSettingsImporter projectSettingsImporter;
+
+  @NotNull
+  private Dialogs dialogs;
+
+  /**
+   * Construct an action with the given main view model provider, project settings importer and
+   * dialog helper.
+   */
+  public ImportProjectSettingsAction(@NotNull MainViewModelProvider mainViewModelProvider,
+                                     @NotNull ProjectSettingsImporter projectSettingsImporter,
+                                     @NotNull Dialogs dialogs) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.projectSettingsImporter = projectSettingsImporter;
+    this.dialogs = dialogs;
+  }
+
+  public ImportProjectSettingsAction() {
+    this(PluginSettings.getInstance(), ImportProjectSettingsAction::doImport,
+        new IntelliJDialogs());
+  }
+
+  @Override
+  public void actionPerformed(@NotNull AnActionEvent e) {
+    Project project = e.getProject();
+    if (project == null) {
+      informNoProjectOpen();
+      return;
+    }
+
+    CourseViewModel courseViewModel = mainViewModelProvider
+        .getMainViewModel(project)
+        .courseViewModel
+        .get();
+    if (courseViewModel == null) {
+      return;
+    }
+    Course course = courseViewModel.getModel();
+
+    URL settingsUrl = course.getResourceUrls().get("projectSettings");
+    if (settingsUrl == null) {
+      informNoSettings(course);
+      return;
+    }
+
+    try {
+      projectSettingsImporter.doImport(project, settingsUrl);
+    } catch (IOException | UnexpectedResponseException ex) {
+      informErrorOccurred();
+    }
+  }
+
+  /**
+   * Downloads the project settings ZIP file from the given URL to a temporary file. After that the
+   * files from the .idea directory of the ZIP file are extracted to the .idea directory of the
+   * given project, after which the project is reloaded.
+   */
+  private static void doImport(@NotNull Project project, @NotNull URL settingsUrl)
+      throws IOException, UnexpectedResponseException {
+    File settingsZip = FileUtilRt.createTempFile(project.getName() + "-settings", ".zip");
+    CoursesClient.fetchZip(settingsUrl, settingsZip);
+    ZipFile zipFile = new ZipFile(settingsZip);
+
+    Path settingsPath = Paths.get(project.getBasePath(), Project.DIRECTORY_STORE_FOLDER);
+
+    List<String> fileNames = zipFile
+        .getFileHeaders()
+        .stream()
+        .filter(file -> !file.isDirectory())
+        .map(FileHeader::getFileName)
+        .collect(Collectors.toList());
+
+    for (String fileName : fileNames) {
+      Path path = Paths.get(fileName);
+      Path pathWithoutRoot = path.subpath(1, path.getNameCount());
+      zipFile.extractFile(path.toString(), settingsPath.toString(), pathWithoutRoot.toString());
+    }
```

Comment:
Nit: this can't last forever, we need to talk... https://www.vavr.io/vavr-docs/ ?

Reply:
Can you elaborate? I understand that `fileNames` could be processed e.g. via Streams API to contain the data in the expected form but that seems at most marginal improvement. Is there more that you see problematic here?

Reply:
the problem is, that `fileNames` actually can't be processed within the same Stream, because I think the `....extractFile(...)` throws an exception. and exceptions can neither be handled in Streams nor be pushed to method signature, so one has to break the Stream and introduce a old-fashioned for-loop to handle the situation. I've been into this several times myself, so that's why the vague comment, sorry. so there are some tools (like Vavr) to work with those cases, but here it's more of a discussion-starter...

Reply:
Excellent discussion starter - I constantly run into this issue of checked exceptions making streams a pain in the ... to use :smile: Luckily in this case the result is still fine. Thanks for the interesting link :+1:

Reply:
For this discussion: #131